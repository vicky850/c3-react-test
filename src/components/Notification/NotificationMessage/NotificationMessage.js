import React from 'react';
import PropTypes from 'prop-types';
import '../../../assets/style/css/NotificationMessage.css';
/**
 * Summary: Notification Message
 * Description: display error, success or warning message based on type 
 * @author Vikash Kumar.
 * @date  11.07.2018
 */
const NotificationMessage = (props) => {
    return (
        <div>
            <div id="alert-message" className={`alert alert-${props.type}`}>{props.message}</div>
        </div>
    );
}
NotificationMessage.propTypes = {
    message: PropTypes.string.isRequired
};

export default NotificationMessage;