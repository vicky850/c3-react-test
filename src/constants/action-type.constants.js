/**
 * Summary: action types
 * Description: action types
 * @author Vikash Kumar.
 * @date  28.06.2018
 */
export const actionTypes = {
    DELETE_DATA: 'delete',
    GET_DATA: 'get'
};
