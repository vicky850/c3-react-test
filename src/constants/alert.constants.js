/**
 * Summary: alert constants
 * Description: alert constants
 * @author Vikash Kumar.
 * @date  28.06.2018
 */
export const alertConstants = {
    SUCCESS: 'success',
    ERROR: 'danger'
};
/*
*message prompts
*/
export const messageConstants = {
    CONFIRM_DELETE: 'Do you want to delete this User?'
};