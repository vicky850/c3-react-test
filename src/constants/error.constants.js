/**
 * Summary: error constants
 * Description: error constants
 * @author Vikash Kumar.
 * @date  28.06.2018
 */
export const customErrorConst = {
    ERROR_DEFAULT: 'Unable to perform this action. Please try later!'
}
export const httpErrorConst = {
    500 : "Internal Server Error."
};
//check valid reponse codes 
export const validCodes=(code)=>{
    if(code>=200 && code<402){
        return true;
    }
    return false;
}
