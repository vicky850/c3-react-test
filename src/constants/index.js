export * from './alert.constants';
export * from './uri.constants';
export * from './error.constants';
export * from './action-type.constants';