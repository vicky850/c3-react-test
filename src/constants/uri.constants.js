/**
 * Summary: define all uri constants
 * Description: define all uri constants
 * @author Vikash Kumar.
 * @date  28.06.2018
 */

export const API_DIRECTORY = { DOC_ROOT: "/api" }

export const urlConstants = {
    BASE_URL: window.BASE_URL + API_DIRECTORY.DOC_ROOT
};
export const API_INTERFACE = {
    GET_USERS: "/users"
}
export const uriConstants = {
    GET_USERS: urlConstants.BASE_URL + API_INTERFACE.GET_USERS
}
