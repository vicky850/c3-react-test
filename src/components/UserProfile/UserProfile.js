/**
 * Summary: User Profile
 * Description: User profile (stateless component)
 * @author Vikash Kumar.
 * @date  28.06.2018
 */
import React from 'react';
import PropTypes from 'prop-types';
import { actionTypes } from '../../constants/action-type.constants';

require('../../assets/style/css/UserProfile.css');

export const UserProfile = (props) => {
  return (
    <div className="box">
      <div className="profile-image"><img src={props.avatar} alt="" /></div>
      <h2>{props.first_name} {props.last_name}</h2>
      <span className="delete" data-action={actionTypes.DELETE_DATA} onClick={(e) => props.optionHandler(e)}>Delete</span>
    </div>
  )
}
/*
*required props
*/
UserProfile.propTypes = {
  first_name: PropTypes.string.isRequired,
  last_name: PropTypes.string.isRequired,
  avatar: PropTypes.string.isRequired
};