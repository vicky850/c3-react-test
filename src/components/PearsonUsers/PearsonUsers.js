/**
 * Summary: PersonUsers component which render the UserProfile list
 * Description: PersonUsers component which render the UserProfile list
 * @author Vikash Kumar.
 * @date  28.06.2018
 */
import React, { Component } from "react";
import { UserProfile } from '../UserProfile/UserProfile'
import { actionTypes } from '../../constants/action-type.constants';
import { alertConstants } from '../../constants/alert.constants';
import { userActions } from '../../actions/user.actions';
import NotificationMessage from '../Notification/NotificationMessage/NotificationMessage';
import ModalDialog from '../Notification/ModalDialog/ModalDialog';
import Loader from '../Loader/Loader';

import _ from 'lodash';
require('../../assets/style/css/PearsonUsers.css');

/*
*Pearson User Class
*/
export class PearsonUsers extends Component {

  /*
  *initialisation of state and bindings
  */
  constructor(props) {
    super(props);
    this.optionHandler = this.optionHandler.bind(this);
    this.state = {
      users: [
        {
          id: 4,
          first_name: "Eve",
          last_name: "Holt",
          avatar:
            "https://s3.amazonaws.com/uifaces/faces/twitter/marcoramires/128.jpg"
        },
        {
          id: 5,
          first_name: "Charles",
          last_name: "Morris",
          avatar:
            "https://s3.amazonaws.com/uifaces/faces/twitter/stephenmoon/128.jpg"
        },
        {
          id: 6,
          first_name: "Tracey",
          last_name: "Ramos",
          avatar:
            "https://s3.amazonaws.com/uifaces/faces/twitter/bigmancho/128.jpg"
        }
      ],
      alert: { type: null, message: null },
      page: 1,
      per_page: 10,
      dataFetched: false
    };
  }

  /*
  *life cycle hook for fetching data/ajax calls
  */
  componentDidMount() {
    this.displayUsers();
  }

/*
* display users
*/
  displayUsers = () => {
    //get user list
    userActions.getUsers(this.state.page, this.state.per_page).then(
      (response) => {
        if (response.type === alertConstants.SUCCESS) {
          const allUsers = [...this.state.users, ...response.data];
          const error_alert = { type: null, message: null };
          this.setState({ users: this.removeDuplicate(allUsers), alert: error_alert, dataFetched: true });
        } else if (response.type === alertConstants.ERROR) {
          const error_alert = { type: response.type, message: response.message.message };
          this.setState({ alert: error_alert, dataFetched: false });
        }
      }
    ).catch(this.errorHandle);
  }

/*
* remove duplicate entries
*/
  removeDuplicate(dataArray) {
    return _.map(_.mapKeys(dataArray, 'id'));
  }

/*
* delete users based on user id
*/
  deleteUser(id) {
    const newUserList = [...this.state.users]; // make a separate copy of the array
    const index = newUserList.findIndex(user => user.id === id);
    newUserList.splice(index, 1);
    this.setState({ users: newUserList });
  }

/*
* user action, delete action here
*/
  optionHandler(e, id) {
    if (e.target.dataset.action === actionTypes.DELETE_DATA) {
      ModalDialog("Are you sure, you want to delete this user ?", () => this.deleteUser(id));
    }
  }

  /*
  *get userprofile component
  * */
  getUserProfile = () => {
    return (
      <div className="bx-wrapper">
        {
          this.state.users.map((user) => {
            return (
              <UserProfile key={user.id} optionHandler={(e) => this.optionHandler(e, user.id)} {...user} />
            );
          })
        }
      </div>
    );
  }

  /*
  *get notification component
  * */
  getNotification = (alert) => {
    return (
      <NotificationMessage {...alert} />
    );
  }

  /*
  *get component based on condition
  * */
  getComponent = () => {
    let component = <Loader />
    component = this.state.alert.message ? this.getNotification(this.state.alert)
      : (this.state.dataFetched ? this.getUserProfile()
        : component);
    return component;
  }

  /*
  *render component
  * */
  render() {
    return (
      <div className="pearson-users">
        <h1>Pearson User Management</h1>
          {this.getComponent()}
      </div>
    );
  }
}
