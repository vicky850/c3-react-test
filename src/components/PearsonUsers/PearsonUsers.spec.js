import React from 'react';
import { configure, shallow } from "enzyme";
import Adapter from 'enzyme-adapter-react-16';
import { PearsonUsers } from "./PearsonUsers";
import { UserProfile } from '../UserProfile/UserProfile';
import NotificationMessage from '../Notification/NotificationMessage/NotificationMessage';
import MockUsers from '../../_mockdata/PearsonUsers.json';
import {customErrorConst, alertConstants} from "../../constants";

configure({ adapter: new Adapter() })
//test case for rendring user list and elements
describe('PearsonUsers Test Suite', function () {
    let component;
    beforeEach(() => {
        component = shallow(<PearsonUsers />);
        component.setState({
            dataFetched: false,
            alert: { type: null, message: null }
        });
    })

    afterEach(() => {
        component = null;
    });

    it('should render without throwing any error', function () {
        expect(component.length).toEqual(1)
    })

    it("renders a h1", () => {
        const h1 = component.find("h1");
        expect(h1.text()).toEqual("Pearson User Management");
    });

    it("it should render a UserProfile", () => {
        component.setState({ dataFetched: true });
        const totalUsers = component.state().users.length;
        expect(component.find(UserProfile)).toHaveLength(totalUsers);
    });

    it("it should render initial layout", () => {
        expect(component.getElements()).toMatchSnapshot();
    });

    it('it should return UserProfile list when data is loaded', () => {
        const spy = jest.spyOn(component.instance(), 'getUserProfile');
        component.setState({ dataFetched: true });
        expect(spy).toHaveBeenCalledTimes(1);
        expect(component.find(UserProfile).exists()).toBeTruthy();
    });

    it("it should remove duplicate entries", () => {
        const totalUsers = [...component.state().users, ...MockUsers];
        const filteredUserList = component.instance().removeDuplicate(totalUsers, 'id');
        const userIds = filteredUserList.map((user) => user.id);
        const ifDuplicate = userIds.some((id, index) => userIds.indexOf(id) != index);
        expect(filteredUserList.length).toBe(3);
        expect(ifDuplicate).toBeFalsy();
    });

    it('it should delete user by its id property', () => {
        component.setState({ users: [...MockUsers] });
        const userId = component.state().users[0].id;
        component.instance().deleteUser(userId);
        const ifUserExist = component.state().users.some((user) => user.id === userId); //check if user exists after deletion
        expect(ifUserExist).toBeFalsy();
    });

    it('it should return Notification Message when data is not presend due to some error', () => {
        const spy = jest.spyOn(component.instance(), 'getNotification');
        component.setState({ alert: { type: alertConstants.ERROR, message: customErrorConst.ERROR_DEFAULT } });
        expect(spy).toHaveBeenCalledTimes(1);
        expect(component.find(NotificationMessage).exists()).toBeTruthy();
    });

})
