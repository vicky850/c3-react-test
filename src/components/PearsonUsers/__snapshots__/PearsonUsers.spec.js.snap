// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PearsonUsers Test Suite it should render initial layout 1`] = `
Array [
  <div
    className="pearon-users"
  >
    <h1>
      Pearson User Management
    </h1>
    <Loader />
  </div>,
]
`;
